libafl_libfuzzer/libafl_libfuzzer_runtime/src/corpus.rs

```diff
@@ -0,0 +1,322 @@
+            current: None,
+            next_recency: AtomicU64::new(0),
+        }
+    }
+
+    pub fn dir_path(&self) -> &PathBuf {
+        &self.corpus_dir
+    }
+
+    /// Touch this index and maybe evict an entry if we have touched an input which was unloaded.
+    fn touch(&self, idx: CorpusId) -> Result<(), Error> {
+        let mut loaded_mapping = self.loaded_mapping.borrow_mut();
+        let mut loaded_entries = self.loaded_entries.borrow_mut();
+        match loaded_mapping.entry(idx) {
+            Entry::Occupied(mut e) => {
+                let &old = e.get();
+                let new = self.next_recency.fetch_add(1, Ordering::Relaxed);
+                e.insert(new);
+                loaded_entries.remove(&old);
+                loaded_entries.insert(new, idx);
+            }
+            Entry::Vacant(e) => {
+                // new entry! send it in
+                let new = self.next_recency.fetch_add(1, Ordering::Relaxed);
+                e.insert(new);
+                loaded_entries.insert(new, idx);
+            }
+        }
+        if loaded_entries.len() > self.max_len {
+            let idx = loaded_entries.pop_first().unwrap().1; // cannot panic
+            let cell = self.mapping.get(idx).ok_or_else(|| {
+                Error::key_not_found(format!("Tried to evict non-existent entry {idx}"))
+            })?;
+            let mut tc = cell.try_borrow_mut()?;
+            let _ = tc.input_mut().take();
+        }
+        Ok(())
+    }
+}
+
+impl<I> UsesInput for LibfuzzerCorpus<I>
+where
+    I: Input + Serialize + for<'de> Deserialize<'de>,
+{
+    type Input = I;
+}
+
+impl<I> Corpus for LibfuzzerCorpus<I>
+where
+    I: Input + Serialize + for<'de> Deserialize<'de>,
+{
+    fn count(&self) -> usize {
+        self.mapping.map.len()
+    }
+
+    fn add(&mut self, testcase: Testcase<Self::Input>) -> Result<CorpusId, Error> {
+        let idx = self.mapping.insert(RefCell::new(testcase));
+        let mut testcase = self.mapping.get(idx).unwrap().borrow_mut();
+
+        match testcase.file_path() {
+            Some(path) if path.canonicalize()?.starts_with(&self.corpus_dir) => {
+                // if it's already in the correct dir, we retain it
+            }
+            _ => {
+                let input = testcase.input().as_ref().ok_or_else(|| {
+                    Error::empty(
+                        "The testcase, when added to the corpus, must have an input present!",
+                    )
+                })?;
+                let name = input.generate_name(idx.into());
+                let path = self.corpus_dir.join(&name);
+
+                match input.to_file(&path) {
+                    Err(Error::File(e, _)) if e.kind() == ErrorKind::AlreadyExists => {
+                        // we do not care if the file already exists; in this case, we assume it is equal
+                    }
+                    res => res?,
+                }
+
+                // we DO NOT save metadata!
```

Comment:
why we don't save metadata?
(maybe because it's never evicted so no need to save it on dthe isk?)

Reply:
There's no point for libafl_libfuzzer -- all of the metadata will be recovered at runtime for effectively no cost.